dynamic_programming/catalan_numbers.nim

```diff
@@ -0,0 +1,105 @@
+# Catalan Numbers
+#[  
+    The Catalan numbers are a sequence of natural numbers that occur in the 
+    most large set of combinatorial problems.
+    For example, it describes:
+    - the number of ways to parenthesize a product of n factors
+    - the number of ways to form a binary search tree with n+1 leaves
+    - the number of Dyck paths of length 2n
+    - the number of ways to triangulate a convex polygon with n+2 sides
+    References:
+    https://en.wikipedia.org/wiki/Catalan_number
+    https://oeis.org/A000108
+]#
+import math
+{.push raises: [].}
+
+func catalanNumbersRecursive(index: Natural): Positive =
+  ## Returns the index-th Catalan number recursively.
+  ## As Nim does not make automatic memoization, this function is not
+  ## efficient.
+  if index == 0:
+    return 1
+  for i in 0 ..< index:
+    result += catalan_numbers_recursive(i) *
+     catalan_numbers_recursive(index - i - 1)
+
+func catalanNumbersRecursive2(index: Natural): Positive =
+  if index == 0:
+    return 1
+  else:
+    2 * (2 * index - 1) * catalanNumbersRecursive2(index - 1) div (index + 1)
+
+func catalanNumbers(index_limit: Natural): seq[Positive] =
+  ## Returns all Catalan numbers up to the index-th number iteratively.
+  var catalan_numbers = newSeq[Positive](index_limit + 1)
+  catalan_numbers[0] = 1
+  for i in 1 .. index_limit:
+    for j in 0 ..< i:
+      catalan_numbers[i] += catalan_numbers[j] * catalan_numbers[i - j - 1]
+  catalan_numbers
+
+func catalanNumbers2(index: Natural): Positive =
+  ## Returns the index-th Catalan number iteratively with a second formula.
+  ## This function is not efficient.
+  binom(2 * index, index) div (index + 1)
+
+iterator catalanNumbersIt(index: Natural): Positive =
+  ## Iterates over all Catalan numbers up to the index-th number.
+  var catalan_numbers_lst = newSeq[Positive](index + 1)
+  catalan_numbers_lst[0] = 1
+  yield 1
+  for i in 1 .. index:
+    for j in 0 ..< i:
+      catalan_numbers_lst[i] += catalan_numbers_lst[j] * catalan_numbers_lst[i - j - 1]
+    yield catalan_numbers_lst[i]
```

Comment:
Multiple yields in regular Nim iterators should be avoided as they can lead to considerable code duplication. Here it's doable and is even shorter.

```suggestion
  var catalanNumbers = newSeq[Positive](index + 1)
  catalanNumbers[0] = 1
  for i in 0 .. index:
    for j in 0 ..< i:
      catalanNumbers[i] += catalanNumbers[j] * catalanNumbers[i - j - 1]
    yield catalanNumbers[i]
```